zone2/main.c

```diff
@@ -102,33 +116,58 @@ void cb_telnet(uint16_t ev, struct pico_socket *s)
         sent_mode = 0;
     }
 
-    if ((ev & PICO_SOCK_EV_RD) && sent_mode) {
-        bytes = pico_socket_read(s, buf, sizeof(buf));
-
-        if (buf[0] == '\xff') {
-            bytes = 0;
-        } else if (buf[0] == '\x0a') {
-            bytes = 2;
-            buf[0] = '\x0d';
-            buf[1] = '\x0a';
-        } else if (buf[0] == '\x0d') {
-            bytes = 2;
-            buf[0] = '\x0d';
-            buf[1] = '\x0a';
+    if ((ev & PICO_SOCK_EV_WR) && !sent_mode) {
+        const char mode[] = IAC DONT LINEMODE
+                            IAC WILL SUPRESS_GO_AHEAD
+                            IAC WILL ECHO;
+
+        pico_socket_write(s, mode, sizeof(mode));
+        sent_mode = 1;
+    } else if (sent_mode) {
+        if (ev & PICO_SOCK_EV_WR) {
+            int msg[4] = {0,0,0,0};
+
+            ECALL_RECV(1, (void*)msg);
+
+            if((msg[CTL] & CTL_DAT) != 0){
+                if(msg[IND] == (msg_out[ACK] + 1)){
+                    char c = msg[DAT];
+                    pico_socket_write(s, &c, 1);
+                    msg_out[CTL] |= CTL_ACK;
+                    msg_out[ACK] = msg[IND];
+                    resend = 1;
+                }
+            }
+
+            if((msg[CTL] & CTL_ACK != 0) & ack_pending){
+                if(msg[ACK] == ack_index){
+                    ack_index++;
+                }
+
+                ack_pending = 0;
+            }
+        }
+
+        if (ev & PICO_SOCK_EV_RD) {
+            if(!ack_pending){
+                bytes = pico_socket_read(s, buf, 1);
+                if (bytes > 0) {
+                    if (buf[0] == '\xff') { // swallow IAC sequences
```

Comment:
This is not the best approach, as IAC sequences could be followed by actual data. We should fetch and interpret these IAC sequences and possibly add negotation.